haddock-api/src/Haddock/Interface/Rename.hs

```diff
@@ -117,60 +107,74 @@ renameInterface _dflags ignoredSymbols renamingEnv warnings iface =
 
 
 -- | The monad does two things for us: it passes around the environment for
--- renaming, and it returns a list of names which couldn't be found in
+-- renaming, and it returns a 'Set' of 'Name's which couldn't be found in
 -- the environment.
-newtype RnM a =
-  RnM { unRn :: (Name -> (Bool, DocName))
-                -- Name lookup function. The 'Bool' indicates that if the name
-                -- was \"found\" in the environment.
-
-             -> (a, [Name] -> [Name])
-                -- Value returned, as well as a difference list of the names not
-                -- found
-      }
-
-instance Monad RnM where
-  m >>= k = RnM $ \lkp -> let (a, out1) = unRn m lkp
-                              (b, out2) = unRn (k a) lkp
-                          in (b, out1 . out2)
-
-instance Functor RnM where
-  fmap f (RnM lkp) = RnM (first f . lkp)
-
-instance Applicative RnM where
-  pure a = RnM (const (a, id))
-  mf <*> mx = RnM $ \lkp -> let (f, out1) = unRn mf lkp
-                                (x, out2) = unRn mx lkp
-                            in (f x, out1 . out2)
+--
+-- However, storing a @'Set' 'Name'@ can result in a huge memory blowup. To
+-- avoid this problem, we also want to have a filtering function that will
+-- prevent names from being recorded if they don't pass the filter.
+newtype RnM a = RnM { unRn :: ReaderT RnMEnv (State (Set Name)) a }
+  deriving newtype (Functor, Applicative, Monad, MonadReader RnMEnv, MonadState (Set Name))
```

Comment:
I suspect that this is potentially much faster, since the implementations in `mtl` have `INLINE` annotations. 

The `State (Set Name)` *could* be a CPS `WriterT` instead - we only do `modify' (Set.insert name)`, which would be equivalent to `tell (Set.singleton name)`.